lib/internal/event_target.js

```diff
@@ -657,32 +518,37 @@ class EventTarget {
    *   capture?: boolean,
    * }} [options]
    */
-  removeEventListener(type, listener, options = kEmptyObject) {
+  removeEventListener(type, callback, options = kEmptyObject) {
     if (!isEventTarget(this))
       throw new ERR_INVALID_THIS('EventTarget');
     if (arguments.length < 2)
       throw new ERR_MISSING_ARGS('type', 'listener');
-    if (!validateEventListener(listener))
+    if (!validateEventListener(callback))
       return;
 
+    const { capture } = flattenMoreOptions(options)
     type = webidl.converters.DOMString(type);
-    const capture = options?.capture === true;
 
-    const root = this[kEvents].get(type);
-    if (root === undefined || root.next === undefined)
-      return;
+    const eventList = this.kTestIndices.get(type)
 
-    let handler = root.next;
-    while (handler !== undefined) {
-      if (handler.same(listener, capture)) {
-        handler.remove();
-        root.size--;
-        if (root.size === 0)
-          this[kEvents].delete(type);
-        this[kRemoveListener](root.size, type, listener, capture);
-        break;
-      }
-      handler = handler.next;
+    if (!eventList) {
+      return
+    }
+
+    const eventListener = ArrayPrototypeFindIndex(eventList, (value) => {
+      return value.type === type && value.capture === capture && value.callback === callback
+    })
+
+    if (eventListener === -1) {
+      return
+    }
+
+    eventList[eventListener].removed = true
+
+    ArrayPrototypeSplice(eventList, eventListener, 1)
```

Comment:
Splice mutates the whole array. Why don't we use Set to make removing faster?

Reply:
reason listed above. EventTarget is poorly designed for optimizations.